tortoise/tortoise.go

```diff
@@ -764,9 +762,29 @@ func (t *turtle) onHareOutput(lid types.LayerID, bid types.BlockID) {
 		return
 	}
 	t.logger.With().Debug("on hare output", lid, bid, log.Bool("empty", bid == types.EmptyBlockID))
-	t.decided[lid] = struct{}{}
-	if bid != types.EmptyBlockID {
-		t.hareOutput[bid] = support
+	previous, exists := t.hareOutput[lid]
+	if !exists || (exists && previous != bid) {
+		t.hareOutput[lid] = bid
+		if lid.Before(t.minprocessed) && t.mode.isVerifying() && withinHdist(t.Config, lid, t.last) {
+			t.logger.With().Info("local opinion changed within hdist",
+				lid,
+				log.Stringer("verified", t.verified),
+				log.Stringer("previous", previous),
+				log.Stringer("new", bid),
+			)
+			t.verifying.resetWeights()
```

Comment:
why is this needed? all the ballots should remain the same, right?

Reply:
all ballots are re-counted based on the "goodness" computed with new opinion, this method resets previously counted weight